api/src/main/java/edu/wpi/first/shuffleboard/api/plugin/Plugin.java

```diff
@@ -0,0 +1,155 @@
+package edu.wpi.first.shuffleboard.api.plugin;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import edu.wpi.first.shuffleboard.api.data.DataType;
+import edu.wpi.first.shuffleboard.api.sources.SourceType;
+import edu.wpi.first.shuffleboard.api.sources.recording.serialization.TypeAdapter;
+import edu.wpi.first.shuffleboard.api.theme.Theme;
+import edu.wpi.first.shuffleboard.api.widget.Widget;
+
+import java.util.List;
+import java.util.Map;
+
+import javafx.beans.property.BooleanProperty;
+import javafx.beans.property.SimpleBooleanProperty;
+
+/**
+ * Common superclass for plugins that can be loaded by the app at startup or during runtime. Subclasses must have
+ * public no-arg constructor or they will not be loaded.
+ */
+public class Plugin {
+
+  private final String groupId;
+  private final String name;
+  private final String version;
+  private final String description;
+  private final BooleanProperty loaded = new SimpleBooleanProperty(this, "loaded", false);
+
+  protected Plugin(String groupId, String name, String version, String description) {
+    this.groupId = groupId;
+    this.name = name;
+    this.version = version;
+    this.description = description;
+  }
+
+  /**
+   * The group ID of this plugin. For example, the stock plugins have the group ID {@code edu.wpi.first.shuffleboard}.
+   * This must combine with {@link #getName()} to create a unique identifier for this plugin.
+   */
+  public final String getGroupId() {
```

Comment:
DataTypes absolutely cannot, because those names are used in the LiveWindow to say the data type (eg `Analog Input`, `Speed Controller`, etc). Allowing duplicate names would at best use the first registered type or at worst use the most recently registered type.

The widgets issue is a UX one. Seeing multiple instances of the same widget name in a list is confusing. It's impossible to know which name is the one you want. Users may also think duplicate names is a bug and not expect to get different widgets for different names. We have an option of displaying them like plugins, with the name preceded by a group ID. But that's a terrible UX:

```
Show as...
  edu.wpi.first.shuffleboard.BooleanBox
  edu.wpi.first.shuffleboard.ToggleSwitch
  third.party.plugin.BooleanBox
  third.party.plugin.ToggleSwitch
  ToggleButton
```